cpp/src/aztec/stdlib/primitives/memory/ram_table.cpp

```diff
@@ -0,0 +1,257 @@
+    , _index_initialized(other._index_initialized)
+    , _length(other._length)
+    , _ram_id(other._ram_id)
+    , _ram_table_generated_in_composer(other._ram_table_generated_in_composer)
+    , _all_entries_written_to_with_constant_index(other._all_entries_written_to_with_constant_index)
+    , _context(other._context)
+{}
+
+/**
+ * @brief Construct a new ram table<Composer>::ram table object
+ *
+ * @tparam Composer
+ * @param other
+ */
+template <typename Composer>
+ram_table<Composer>::ram_table(ram_table&& other)
+    : _raw_entries(other._raw_entries)
+    , _index_initialized(other._index_initialized)
+    , _length(other._length)
+    , _ram_id(other._ram_id)
+    , _ram_table_generated_in_composer(other._ram_table_generated_in_composer)
+    , _all_entries_written_to_with_constant_index(other._all_entries_written_to_with_constant_index)
+    , _context(other._context)
+{}
+
+/**
+ * @brief Copy assignment operator
+ *
+ * @tparam Composer
+ * @param other
+ * @return ram_table<Composer>&
+ */
+template <typename Composer> ram_table<Composer>& ram_table<Composer>::operator=(const ram_table& other)
+{
+    _raw_entries = other._raw_entries;
+    _length = other._length;
+    _ram_id = other._ram_id;
+    _index_initialized = other._index_initialized;
+    _ram_table_generated_in_composer = other._ram_table_generated_in_composer;
+    _all_entries_written_to_with_constant_index = other._all_entries_written_to_with_constant_index;
+
+    _context = other._context;
+    return *this;
+}
+
+/**
+ * @brief Move assignment operator
+ *
+ * @tparam Composer
+ * @param other
+ * @return ram_table<Composer>&
+ */
+template <typename Composer> ram_table<Composer>& ram_table<Composer>::operator=(ram_table&& other)
+{
+    _raw_entries = other._raw_entries;
+    _length = other._length;
+    _ram_id = other._ram_id;
+    _index_initialized = other._index_initialized;
+    _ram_table_generated_in_composer = other._ram_table_generated_in_composer;
+    _all_entries_written_to_with_constant_index = other._all_entries_written_to_with_constant_index;
+    _context = other._context;
+    return *this;
+}
+
+/**
+ * @brief Read a field element from the RAM table at an index value
+ *
+ * @tparam Composer
+ * @param index
+ * @return field_t<Composer>
+ */
+template <typename Composer> field_t<Composer> ram_table<Composer>::read(const field_pt& index) const
+{
+    if (_context == nullptr) {
+        _context = index.get_context();
+    }
+
+    if (uint256_t(index.get_value()) >= _length) {
+        // TODO: what's best practise here? We are assuming that this action will generate failing constraints,
+        // and we set failure message here so that it better describes the point of failure.
+        // However, we are not *ensuring* that failing constraints are generated at the point that `failure()` is
+        // called. Is this ok?
+        _context->failure("ram_table: RAM array access out of bounds");
+    }
+
+    initialize_table();
```

Comment:
This will fail if all operations were with constants. If we check, whether the context is a nullptr, we can simply return a constant value in the constant case

Reply:
Checking the context pointer isn't sufficient to determine whether the operations are all over constants. Other methods in this class will simply break if the pointer is null (see the asserts in other methods)

It's a major pain to get the RAM table implementation to a state where it can correctly handle pure constant cases. I took the simpler approach where we just instantiate all constants as witnesses and constrain them to be equal to the required values. Will require a substantial rewrite to provide a constant-friendly code path that can accomodate a null context pointer.